File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos33/Alg33.py
'''
♦ Algoritmo 33: Estrutura de Dados Não Linear Dinâmica (não Estática) de Árvore Binária: Árvore de Busca Binária (Binary Search Tree - BST).
'''


class No:
    def __init__(self, valor):
        self.valor = valor
        self.esquerda = None
        self.direita = None

class ArvoreBinaria:
    def __init__(self):
        self.raiz = None

    def inserir(self, valor):
        self.raiz = self.inserir_recursivo(self.raiz, valor)

    def inserir_recursivo(self, no, valor):
        if no is None:
            return No(valor)

        if valor < no.valor:
            no.esquerda = self.inserir_recursivo(no.esquerda, valor)
        elif valor > no.valor:
            no.direita = self.inserir_recursivo(no.direita, valor)

        return no

    def buscar(self, valor):
        return self.buscar_recursivo(self.raiz, valor)

    def buscar_recursivo(self, no, valor):
        if no is None:
            return False

        if valor == no.valor:
            return True

        return self.buscar_recursivo(no.esquerda, valor) if valor < no.valor else self.buscar_recursivo(no.direita, valor)

    def imprimir_em_ordem(self):
        self.imprimir_em_ordem_recursivo(self.raiz)

    def imprimir_em_ordem_recursivo(self, no):
        if no is not None:
            self.imprimir_em_ordem_recursivo(no.esquerda)
            print(no.valor, end=" ")
            self.imprimir_em_ordem_recursivo(no.direita)

arvore = ArvoreBinaria()
arvore.inserir(10)
arvore.inserir(5)
arvore.inserir(20)
arvore.inserir(3)
arvore.inserir(8)
arvore.inserir(15)

print("Em Ordem:")
arvore.imprimir_em_ordem()

print("\n\nBuscar 8:", arvore.buscar(8))
print("Buscar 12:", arvore.buscar(12))

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos33/Alg33.cpp
/*
♦ Algoritmo 33: Estrutura de Dados Não Linear Dinâmica (não Estática) de Árvore Binária: Árvore de Busca Binária (Binary Search Tree - BST).
*/


#include <iostream>

template <typename T>
struct No {
    T valor;
    No<T>* esquerda;
    No<T>* direita;

    No(T valor) : valor(valor), esquerda(nullptr), direita(nullptr) {}
};

template <typename T>
class ArvoreBinaria {
private:
    No<T>* raiz;

    No<T>* inserirRecursivo(No<T>* no, T valor) {
        if (no == nullptr) {
            return new No<T>(valor);
        }

        if (valor < no->valor) {
            no->esquerda = inserirRecursivo(no->esquerda, valor);
        } else if (valor > no->valor) {
            no->direita = inserirRecursivo(no->direita, valor);
        }

        return no;
    }

    bool buscarRecursivo(No<T>* no, T valor) {
        if (no == nullptr) {
            return false;
        }

        if (valor == no->valor) {
            return true;
        }

        return valor < no->valor ? buscarRecursivo(no->esquerda, valor) : buscarRecursivo(no->direita, valor);
    }

    void imprimirEmOrdemRecursivo(No<T>* no) {
        if (no != nullptr) {
            imprimirEmOrdemRecursivo(no->esquerda);
            std::cout << no->valor << " ";
            imprimirEmOrdemRecursivo(no->direita);
        }
    }

public:
    ArvoreBinaria() : raiz(nullptr) {}

    void inserir(T valor) {
        raiz = inserirRecursivo(raiz, valor);
    }

    bool buscar(T valor) {
        return buscarRecursivo(raiz, valor);
    }

    void imprimirEmOrdem() {
        imprimirEmOrdemRecursivo(raiz);
    }
};

int main() {
    ArvoreBinaria<int> arvore;
    arvore.inserir(10);
    arvore.inserir(5);
    arvore.inserir(20);
    arvore.inserir(3);
    arvore.inserir(8);
    arvore.inserir(15);

    std::cout << "Em Ordem:\n";
    arvore.imprimirEmOrdem();

    std::cout << "\nBuscar 8: " << std::boolalpha << arvore.buscar(8) << std::endl;
    std::cout << "Buscar 12: " << std::boolalpha << arvore.buscar(12) << std::endl;

    return 0;
}
